helix-core/src/indent.rs

```diff
@@ -849,24 +936,76 @@ pub fn indent_for_newline(
     current_line: usize,
 ) -> String {
     let indent_width = indent_style.indent_width(tab_width);
-    if let (Some(query), Some(syntax)) = (
+    if let (
+        IndentationHeuristic::TreeSitter | IndentationHeuristic::Hybrid,
+        Some(query),
+        Some(syntax),
+    ) = (
+        indent_heuristic,
         language_config.and_then(|config| config.indent_query()),
         syntax,
     ) {
         if let Some(indent) = treesitter_indent_for_pos(
             query,
             syntax,
-            indent_style,
             tab_width,
             indent_width,
             text,
             line_before,
             line_before_end_pos,
             true,
         ) {
-            return indent;
+            if let IndentationHeuristic::Hybrid = indent_heuristic {
+                // We want to compute the indentation not only based on the
+                // syntax tree but also on the actual indentation of a previous
+                // line. This makes indentation computation more resilient to
+                // incomplete queries, incomplete source code & differing indentation
+                // styles for the same language.
+                // However, using the indent of a previous line as a baseline may not
+                // make sense, e.g. if it has a different alignment than the new line.
+                // In order to prevent edge cases with long running times, we only try
+                // a constant number of (non-empty) lines.
+                const MAX_ATTEMPTS: usize = 2;
+                let mut num_attempts = 0;
+                for line_idx in (0..=line_before).rev() {
```

Comment:
Instead of iterating line indecies you can use `text.lines_at(line_before + 1).reversed()` to iterate lines (it's faster/more ideomatic)

Reply:
You could also handle the max attempt by adding `take(MAX_ATTEMPTS)`

Reply:
Just iterating lines doesn't work here because `treesitter_indent_for_pos` needs a line index. Using `take(MAX_ATTEMPTS)` would work but it would slightly change the logic: Currently, blank lines are not counted towards the number of attempts, since we don't do any indent computation on them. If we did count blank lines, this would cause us to fallback to the `tree-sitter` heuristic much more often.

Reply:
Fair enough, it's a bit slower to iterate lines this way (about 5x) but it doesn't matter too much here I supposed.